File: src/core/effect-executor.ts
import type { Effect, World, ToolCall, ComponentDataMap } from '#src/core/types';
import type { EventBus } from '#src/core/event-bus';
import type { ToolScheduler } from '#src/core/tool-scheduler';
import type { AbortManager } from '#src/core/abort-manager';

export class EffectExecutor {
  constructor(
    private world: World,
    private eventBus: EventBus,
    private toolScheduler: ToolScheduler,
    private abortManager: AbortManager,
  ) {}

  async execute(effects: Effect[]): Promise<void> {
    for (const effect of effects) {
      await this.apply(effect);
    }
  }

  private async apply(effect: Effect): Promise<void> {
    switch (effect.type) {
      case 'effect.callLLM': {
        const provider = this.world.getComponent(effect.provider, 'provider');
        if (!provider) break;
        try {
          const body = provider.formatMessages(effect.messages);
          const stream = provider.streamChat(body);
          for await (const event of stream) {
            this.eventBus.emit(event);
          }
        } catch (err) {
          this.eventBus.emit({ type: 'llm.stream_error', error: String(err) });
        }
        break;
      }
      case 'effect.executeTool': {
        const toolCall: ToolCall = {
          id: `tool_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`,
          type: 'function',
          function: {
            name: effect.tool_name,
            arguments: JSON.stringify(effect.args),
          },
        };
        const results = await this.toolScheduler.schedule([toolCall]);
        const toolResult = results[0];
        if (!toolResult) break;
        if (toolResult.error) {
          this.eventBus.emit({ type: 'tool.error', tool_call_id: toolResult.tool_call_id, error: toolResult.error });
        } else {
          this.eventBus.emit({ type: 'tool.result', tool_call_id: toolResult.tool_call_id, result: toolResult.result! });
        }
        break;
      }
      case 'effect.appendMessage': {
        const comp = this.world.getComponent(effect.entity, 'messages');
        if (comp) {
          comp.messages.push(effect.message);
        }
        break;
      }
      case 'effect.updateComponent':
        this.world.setComponent(
          effect.entity,
          effect.component as keyof ComponentDataMap,
          effect.data as ComponentDataMap[keyof ComponentDataMap],
        );
        break;
      case 'effect.emit':
        this.eventBus.emit(effect.event);
        break;
      case 'effect.render':
        if (effect.stream === 'thinking') {
          process.stdout.write(`\x1b[2m${effect.content}\x1b[0m`);
        } else {
          process.stdout.write(effect.content);
        }
        break;
      case 'effect.abort':
        this.abortManager.abort(effect.reason);
        break;
      default: {
        const _exhaustive: never = effect;
        break;
      }
    }
  }
}

File: crates/alice-core/src/effect_executor.rs
use crate::components::{
    ComponentAccessor, ConfigComponent, LoopComponent, MessagesComponent, ProviderComponent,
    ToolsComponent,
};
use crate::effect::Effect;
use crate::event::{Event, ToolEvent};
use crate::event_bus::EventBus;
use crate::tool_scheduler::ToolScheduler;
use crate::abort_manager::AbortManager;
use crate::types::{FunctionCall, ToolCall};
use crate::world::{HasComponent, World};

pub struct EffectExecutor<'a, Components> {
    world: &'a mut World<Components>,
    event_bus: &'a EventBus,
    tool_scheduler: &'a ToolScheduler,
    abort_manager: &'a mut AbortManager,
}

impl<'a, Components> EffectExecutor<'a, Components>
where
    Components: HasComponent<MessagesComponent>
        + HasComponent<LoopComponent>
        + HasComponent<ConfigComponent>
        + HasComponent<ToolsComponent>
        + HasComponent<ProviderComponent>,
    World<Components>: ComponentAccessor,
{
    pub fn new(
        world: &'a mut World<Components>,
        event_bus: &'a EventBus,
        tool_scheduler: &'a ToolScheduler,
        abort_manager: &'a mut AbortManager,
    ) -> Self {
        Self {
            world,
            event_bus,
            tool_scheduler,
            abort_manager,
        }
    }

    pub fn execute(&mut self, effects: Vec<Effect>) {
        for effect in effects {
            self.apply(effect);
        }
    }

    fn apply(&mut self, effect: Effect) {
        match effect {
            Effect::AppendMessage { entity: _, message } => {
                self.world.get_mut::<MessagesComponent>().messages.push(message);
            }
            Effect::UpdateComponent { entity: _, update } => {
                update.apply(self.world);
            }
            Effect::Emit { event } => {
                self.event_bus.emit(&event);
            }
            Effect::Render { content, stream: _ } => {
                print!("{}", content);
            }
            Effect::Abort { reason } => {
                self.abort_manager.abort(reason);
            }
            Effect::ExecuteTool { tool_name, args } => {
                let tool_call = ToolCall {
                    id: format!("tool_{}", uuid_simple()),
                    call_type: "function".into(),
                    function: FunctionCall {
                        name: tool_name,
                        arguments: serde_json::to_string(&args).unwrap_or_default(),
                    },
                };
                let results = self.tool_scheduler.schedule(&[tool_call]);
                if let Some(r) = results.into_iter().next() {
                    match r.error {
                        Some(err) => {
                            self.event_bus.emit(&Event::Tool(ToolEvent::Error {
                                tool_call_id: r.tool_call_id,
                                error: err,
                            }));
                        }
                        None => {
                            self.event_bus.emit(&Event::Tool(ToolEvent::Result {
                                tool_call_id: r.tool_call_id,
                                result: r.result.unwrap_or_default(),
                            }));
                        }
                    }
                }
            }
            Effect::CallLLM { messages: _ } => {
                // Placeholder: wired when StreamingProvider is connected in later batch.
            }
        }
    }
}

fn uuid_simple() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    format!("{:x}", ts)
}

impl<T> ComponentAccessor for World<T>
where
    T: HasComponent<MessagesComponent>
        + HasComponent<LoopComponent>
        + HasComponent<ConfigComponent>
        + HasComponent<ToolsComponent>
        + HasComponent<ProviderComponent>,
{
    fn messages_mut(&mut self) -> &mut MessagesComponent {
        self.get_mut::<MessagesComponent>()
    }

    fn config_mut(&mut self) -> &mut ConfigComponent {
        self.get_mut::<ConfigComponent>()
    }

    fn loop_mut(&mut self) -> &mut LoopComponent {
        self.get_mut::<LoopComponent>()
    }

    fn tools_mut(&mut self) -> &mut ToolsComponent {
        self.get_mut::<ToolsComponent>()
    }

    fn provider_mut(&mut self) -> &mut ProviderComponent {
        self.get_mut::<ProviderComponent>()
    }
}
